locator/abstract_replication_strategy.hh

```diff
@@ -394,7 +396,15 @@ public:
     bool stopped() const noexcept {
         return _stopped;
     }
-
+    template <typename Func>
+    requires requires (Func func, effective_replication_map_ptr erm) {
+        { futurize_invoke(func, erm) } -> std::same_as<future<>>;
+    }
```

Comment:
It's enough to write std::invocable<effective_replication_map_ptr> IMO. Writing code in the concept means that the reader has to parse the code and understand it, whereas the named concept says: give me a function that accepts an effective_replication_map_ptr.

btw, isn't it nicer to accept const effective_replication_map&? the callee probably doesn't care for the pointer.

Reply:
(then it simplfies to a single line: `future<> do_for_each(std::invocable<...> auto func) ... {`)

Reply:
makes sense, will do in next version

Reply:
> btw, isn't it nicer to accept const effective_replication_map&? the callee probably doesn't care for the pointer.

It's nicer, but if the callee is async it needs to keep the shard pointer to extend its life time.
It can call map.shared_from_this() if needed. I'll document that for future users.

Reply:
Hmm, this doesn't work well:
```
    template <typename Func>
    future<> do_for_each(std::invocable<Func, const effective_replication_map&> auto func) const noexcept {
```

```
service/storage_service.cc:1872:34: error: no matching member function for call to 'do_for_each'
            ss.get_erm_factory().do_for_each([endpoint] (const locator::effective_replication_map& erm) {
            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
./locator/abstract_replication_strategy.hh:407:14: note: candidate template ignored: couldn't infer template argument 'Func'
    future<> do_for_each(std::invocable<Func, const effective_replication_map&> auto func) const noexcept {
             ^
```

while this does work:
```
    template <typename Func>
    requires std::invocable<Func, const effective_replication_map&>
    future<> do_for_each(Func func) const noexcept {
```

Reply:
Drop `template <typename Func>`

Reply:
All three lines drop to one:

```c++
future<> do_for_each(std::invocable<const effective_replication_map&> auto func) const noexcept {
```

Reply:
Fixed in v4